enhancements/network/ip-interface-selection.yaml

```diff
@@ -0,0 +1,302 @@
+---
+title: ip-interface-selection
+authors:
+  - @cybertron
+reviewers:
+  - @jcaamano
+  - @tsorya
+approvers:
+  - TBD
+api-approvers:
+  - "None"
+creation-date: 2022-07-07
+last-updated: 2022-07-07
+tracking-link: # link to the tracking ticket (for example: Jira Feature or Epic ticket) that corresponds to this enhancement
+  - TBD
+see-also:
+  - https://github.com/openshift/baremetal-runtimecfg/issues/119
+replaces:
+superseded-by:
+---
+
+# IP and Interface Selection
+
+## Summary
+
+As OpenShift is deployed in increasingly complex networking environments, we
+have gotten many requests for more control over which interface is used for
+the primary node IP. We provided a basic mechanism for this with
+[KUBELET_NODEIP_HINT](https://github.com/openshift/machine-config-operator/pull/2888)
+but as users have started to exercise that some significant limitations have
+come to light.
+
+## Motivation
+
+Some users want to have a great deal of control over how their network traffic
+is routed. Because we use the default route for interface and IP selection,
+in some cases they are not able to route traffic the way they want.
+
+### User Stories
+
+As a network administrator, I want cluster traffic to be on one NIC and the
+default route on a different NIC so external traffic is segregated from
+internal traffic.
+
+TODO: I'd like to get a more specific user story since I know we have a
+number of requests for this.
+
+### Goals
+
+Ensure that all host networked services on a node have consistent interface
+and IP selection.
+
+### Non-Goals
+
+Support for platforms other than UPI (None).
+
+Complete support for multiple nics, with full control over what traffic
+gets routed where. However, this work should be able to serve as the basis
+for a broader multi-nic feature so we should avoid designs that would limit
+future work in this area.
+
+## Proposal
+
+The following is a possibly incomplete list of places that we do IP/interface
+selection. Ideally these should all use a single mechanism to do that so they
+are all consistent with each other.
+
+- Node IP (Kubelet and CRIO)
+- configure-ovs
+- resolv-prepender
+- Keepalived
+- Etcd
+
+In some cases (local DNS) it may not matter if the IP selected is consistent
+with the other cases, but Node IP, configure-ovs, and Keepalived all need to
+match because their functionality depends on it. I'm less familiar with the
+requirements for Etcd, but it seems likely that should match as well. In
```

Comment:
@rbbratta Had run into issues with etcd when the master nodes selected the wrong interface, so etcd is definitely some thing to consider.

Reply:
I tried to track it down once, but not an etcd expert.

https://bugzilla.redhat.com/show_bug.cgi?id=2057160#c24

cluest-etcd-operator uses `GetPreferredInternalIPAddressForNodeName` which iterates over `node.Status.Addresses`?

https://github.com/openshift/cluster-etcd-operator/blob/master/pkg/dnshelpers/util.go#L40

```go
	for _, currAddress := range node.Status.Addresses {
		if currAddress.Type == corev1.NodeInternalIP {
			switch ipFamily {
			case "tcp4":
```

`node.Status.Addresses` is not sorted?

Is looks like kubelet `nodeAddressesFunc` provides node.Status.Addresses

https://github.com/openshift/origin/blob/master/vendor/k8s.io/kubernetes/pkg/kubelet/kubelet_node_status.go#L642

```go
	// if cloud is not nil, we expect the cloud resource sync manager to exist
	var nodeAddressesFunc func() ([]v1.NodeAddress, error)
	if kl.cloud != nil {
		nodeAddressesFunc = kl.cloudResourceSyncManager.NodeAddresses
	}
```